### ISSUE TITLE
请教代码中问题

### ISSUE THREAD

**Issue by XiaosuWang** [microsoft/LiST#13]
https://github.com/microsoft/LiST/blob/4d461bc6b1c910b5886c146ca01dbb3440ab7a3b/src/trainer.py#L1335-L1341

尊敬的作者，您好，请问上面这段代码中，在warmup student 的同时，为什么也要同时在有标签的数据上train student，这样混合的训练，与论文中的算法流程好像有别？请问是我的理解错了，还是确实代码有问题？如果代码有问题，请问如何修改？谢谢

**Comment by yaqingwang** [comment]
Hi Xiaosu, thanks for your interest in our project. I am not very sure if i understand you question correctly but I will try to explain and please let me know if this answers your question. First, during the warmup stage, we only use unlabeled data. We do have a procedure of iteratively training student with unlabeled data and labeled data. The details can be found in algorithm flow, located in appendix algorithm 1 ( https://arxiv.org/pdf/2110.06274.pdf).    Finetune parameter is used to control switch between these two stages.

**Comment by XiaosuWang** [comment]
@yaqingwang thank you for your answer, I have understood the source code about the above question. Another question is that, after warmup student, which kind of label of the unlabeled data is used to train student, soft label or hard label? As shown in the below lines, after warmup student, the value of use_soft_label is 1 or not?
https://github.com/microsoft/LiST/blob/4d461bc6b1c910b5886c146ca01dbb3440ab7a3b/src/trainer.py#L633-L637

**Comment by yaqingwang** [comment]
Glad to know this answers your question. For hard label or soft label, both of options are open and actually up to you. In our paper, we prefer hard labels, which work well with the proposed re-weighting mechanism in the paper. But please be aware that hard labels and soft labels have some trade-offs. If there is a good re-weighting mechanism, hard labels bring more gains, otherwise,  soft labels are generally considered to be a safe (or conservative) choice.

**Comment by XiaosuWang** [comment]
@yaqingwang thank you very much, my confusion has been resolved.
